hardinfo/dmi_util.c

```diff
@@ -0,0 +1,152 @@
+/*
+ *    HardInfo - Displays System Information
+ *    Copyright (C) 2003-2017 Leandro A. F. Pereira <leandro@hardinfo.org>
+ *    This file
+ *    Copyright (C) 2017 Burt P. <pburt0@gmail.com>
+ *
+ *    This program is free software; you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, version 2.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with this program; if not, write to the Free Software
+ *    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "hardinfo.h"
+#include "dmi_util.h"
+
+const char *dmi_sysfs_root() {
+    char *candidates[] = {
+        "/sys/devices/virtual/dmi",
+        "/sys/class/dmi",
+        NULL
+    };
+    int i = 0;
+    while (candidates[i] != NULL) {
+        if(access(candidates[i], F_OK) != -1)
+            return candidates[i];
+        i++;
+    }
+    return NULL;
+}
+
+char *dmi_get_str(const char *id_str) {
+  static struct {
+    char *id;
+    char *path;
+  } tab_dmi_sysfs[] = {
+    /* dmidecode -> sysfs */
+    { "bios-release-date", "id/bios_date" },
+    { "bios-vendor", "id/bios_vendor" },
+    { "bios-version", "id/bios_version" },
+    { "baseboard-product-name", "id/board_name" },
+    { "baseboard-manufacturer", "id/board_vendor" },
+    { "baseboard-version", "id/board_version" },
+    { "baseboard-serial-number", "id/board_serial" },
+    { "baseboard-asset-tag", "id/board_asset_tag" },
+    { "system-product-name", "id/product_name" },
+    { "system-manufacturer", "id/sys_vendor" },
+    { "system-serial-number", "id/product_serial" },
+    { "system-product-family", "id/product_family" },
+    { "system-version", "id/product_version" },
+    { "system-uuid", "product_uuid" },
+    { "chassis-type", "id/chassis_type" },
+    { "chassis-serial-number", "id/chassis_serial" },
+    { "chassis-manufacturer", "id/chassis_vendor" },
+    { "chassis-version", "id/chassis_version" },
+    { "chassis-asset-tag", "id/chassis_asset_tag" },
+    { NULL, NULL }
+  };
+  const gchar *dmi_root = dmi_sysfs_root();
+  gchar *full_path = NULL, *ret = NULL;
+  gboolean spawned;
+  gchar *out, *err;
+
+  int i = 0;
+
+  /* try sysfs first */
+  if (dmi_root) {
+    while (tab_dmi_sysfs[i].id != NULL) {
+      if (strcmp(id_str, tab_dmi_sysfs[i].id) == 0) {
+        full_path = g_strdup_printf("%s/%s", dmi_root, tab_dmi_sysfs[i].path);
```

Comment:
Unless you need to store a path, always declare it as a array on the stack.  This reduces heap memory allocation, and it makes it easier to spot leaks (since it can't leak).  It can be `PATH_MAX` size and you can use `snprintf()` to populate it.